File: Python/Binary Search/M-875. Koko Eating Bananas.py
import math


class Solution:
    @classmethod
    def minEatingSpeed(self, piles: list[int], h: int) -> int:
        min_speed, max_speed = 1, max(piles)
        best_speed = max_speed

        def calc_needed_time(speed: int):
            ttl_time = 0
            for pile in piles:
                ttl_time += math.ceil(pile / speed)
            return ttl_time

        while min_speed <= max_speed:
            speed = (min_speed + max_speed) // 2
            needed_time = calc_needed_time(speed)
            if needed_time <= h:
                best_speed = speed
                max_speed = speed - 1
            else:
                min_speed = speed + 1
        return best_speed


testData = [
    {
        "piles": [3, 6, 7, 11],
        "h": 8,
        "ans": 4,
    },
    {
        "piles": [30, 11, 23, 4, 20],
        "h": 5,
        "ans": 30,
    },
    {
        "piles": [30, 11, 23, 4, 20],
        "h": 6,
        "ans": 23,
    },
]

for data in testData:
    print(data["ans"] == Solution.minEatingSpeed(data["piles"], data["h"]))

File: Binary Search/M-875. Koko Eating Bananas/M-875. Koko Eating Bananas.ts
function minEatingSpeed(piles: number[], h: number): number {
  let [minSpeed, maxSpeed] = [0, Math.max(...piles)];
  let bestSpeed = maxSpeed;
  const calcNeededTime = (speed) =>
    piles.reduce((ttlTime, pile) => {
      return ttlTime + Math.ceil(pile / speed);
    }, 0);

  while (minSpeed <= maxSpeed) {
    const speed = Math.floor((minSpeed + maxSpeed) / 2);
    const neededTime = calcNeededTime(speed);
    if (neededTime <= h) {
      bestSpeed = speed;
      maxSpeed = speed - 1;
    } else {
      minSpeed = speed + 1;
    }
  }
  return bestSpeed;
}

const testData = [
  {
    piles: [3, 6, 7, 11],
    h: 8,
    ans: 4,
  },
  {
    piles: [30, 11, 23, 4, 20],
    h: 5,
    ans: 30,
  },
  {
    piles: [30, 11, 23, 4, 20],
    h: 6,
    ans: 23,
  },
];

for (const data of testData) {
  console.log(data.ans === minEatingSpeed(data.piles, data.h));
}
